### PR TITLE
Instrument Sidekiq traces with Delayed Extensions

### PR DESCRIPTION
**Pull request by marcotc**
_Follow up from #798_

Sidekiq provides [Delayed-extensions](https://github.com/mperham/sidekiq/wiki/Delayed-extensions/c03d6d910ebb5dd7984525beac353d26b41c4ad7) to allow for any method call to be made asynchronous.

When tracing jobs scheduled with Delayed-extensions, we receive information about what class and method were invoked as a YAML argument to the job, instead of the default `job['class']` attribute: `job['class']` is always set to `Sidekiq::Extensions::DelayedClass` for delayed jobs.

This PR decodes that YAML and retrieves information about the job invoked, setting it as the trace resource.

We use `YAML.parse` (instead of `YAML.load`) as it only parses the source text into a syntax tree, without loading any native Ruby objects. `YAML.load` can lead to vulnerabilities during object loading, but `YAML.parse` cannot.

### CODE REVIEW

File: lib/ddtrace/contrib/sidekiq/tracing.rb [hunk 1]
```diff
@@ -18,9 +20,24 @@ def initialize(options = {})
         def job_resource(job)
           if job['wrapped']
             job['wrapped']
+          elsif job['class'] == 'Sidekiq::Extensions::DelayedClass'
+            delay_extension_class(job)
           else
             job['class']
           end
+        rescue => e
+          Datadog::Logger.log.debug("Error retrieving Sidekiq job class name (jid:#{job['jid']}): #{e}")
```
Comment (delner) [inline lib/ddtrace/contrib/sidekiq/tracing.rb]: Side thought, but if this happened on the regular, then user's applications would likely see slowdowns with the raised and handled exceptions, but not have the errors be apparent until they activated debug mode. I don't think there's much to do about it, and overall this is the safest option, though.

File: lib/ddtrace/contrib/sidekiq/tracing.rb [hunk 2]
```diff
@@ -18,9 +20,24 @@ def initialize(options = {})
...
+
+          job['class']
+        end
+
+        #
+        def delay_extension_class(job)
+          clazz, method = YAML.parse(job['args'].first).children.first.children
```
Comment (delner) [inline lib/ddtrace/contrib/sidekiq/tracing.rb]: If `parse` is safe, I think that will address the security concerns from the previous version of this.

File: test/contrib/sidekiq/client_tracer_test.rb [hunk 3]
```diff
@@ -56,4 +61,9 @@ def test_empty_parentless
     assert_equal(0, span.status)
     assert_nil(span.parent)
   end
+
+  def test_delayed_extensions
```
Comment (delner) [inline test/contrib/sidekiq/client_tracer_test.rb]: Would be great to convert these to RSpec, but its not a blocker for this task (out of scope.)

File: lib/ddtrace/contrib/sidekiq/tracing.rb (hunk 1 above)
Comment (marcotc) [reply lib/ddtrace/contrib/sidekiq/tracing.rb]: I'm able to pass the log message as a lazily evaluated block.
I ran some benchmarks and it seems like it's almost 3x as fast when log level is disabled for the message we want to log (the tracer defaults to WARN level, but we are logging a DEBUG message).

I've applied these changes to this PR.

```ruby
Benchmark.ips do |x|
  x.report('block') do
    log.debug { "Error retrieving Sidekiq job class name (jid:#{job['jid']}): #{e}" }
  end
  
  x.report('immediate') do
    log.debug("Error retrieving Sidekiq job class name (jid:#{job['jid']}): #{e}")
  end
end
```

```
Warming up --------------------------------------
               block   307.595k i/100ms
           immediate   127.102k i/100ms
Calculating -------------------------------------
               block      6.441M (±17.4%) i/s -     31.067M in   5.016951s
           immediate      1.780M (± 4.1%) i/s -      8.897M in   5.007571s
```

File: lib/ddtrace/contrib/sidekiq/tracing.rb (hunk 2 above)
Comment (marcotc) [reply lib/ddtrace/contrib/sidekiq/tracing.rb]: 👍

File: test/contrib/sidekiq/client_tracer_test.rb (hunk 3 above)
Comment (marcotc) [reply test/contrib/sidekiq/client_tracer_test.rb]: Agreed!

File: lib/ddtrace/contrib/sidekiq/tracing.rb (hunk 1 above)
Comment (delner) [reply lib/ddtrace/contrib/sidekiq/tracing.rb]: Oh okay fair enough, that wasn't actually what I was pointing to but that works as well. My thought was more about the nature of an exception being repeatedly raised itself (which is expensive), regardless of how its handled. There's no action required on that original thought but your change is fine; thanks!

**Comment by marcotc** [comment]
Thanks to @joeyAghion for the initial implementation!

**Pull request merged**